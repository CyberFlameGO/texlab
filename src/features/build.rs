use std::{
    io::{BufRead, BufReader, Read},
    path::Path,
    process::{Command, Stdio},
    sync::{Arc, Mutex},
    thread::{self, JoinHandle},
};

use anyhow::Result;
use crossbeam_channel::{Receiver, Sender};
use dashmap::DashMap;
use encoding_rs_io::DecodeReaderBytesBuilder;
use lsp_types::{
    notification::{LogMessage, Progress},
    LogMessageParams, NumberOrString, Position, ProgressParams, ProgressParamsValue,
    TextDocumentIdentifier, Url, WorkDoneProgress, WorkDoneProgressBegin,
    WorkDoneProgressCreateParams, WorkDoneProgressEnd,
};
use serde::{Deserialize, Serialize};
use serde_repr::{Deserialize_repr, Serialize_repr};
use uuid::Uuid;

use crate::{client::LspClient, ClientCapabilitiesExt};

use super::{FeatureRequest, ForwardSearch};

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BuildParams {
    pub text_document: TextDocumentIdentifier,
}

#[derive(Debug, PartialEq, Eq, Clone, Copy, Serialize_repr, Deserialize_repr)]
#[repr(i32)]
pub enum BuildStatus {
    SUCCESS = 0,
    ERROR = 1,
    FAILURE = 2,
    CANCELLED = 3,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BuildResult {
    pub status: BuildStatus,
}

struct ProgressReporter<'a> {
    supports_progress: bool,
    client: LspClient,
    token: &'a str,
}

impl<'a> ProgressReporter<'a> {
    pub fn start(&self, uri: &Url) -> Result<()> {
        if self.supports_progress {
            self.client
                .send_request::<lsp_types::request::WorkDoneProgressCreate>(
                    WorkDoneProgressCreateParams {
                        token: NumberOrString::String(self.token.to_string()),
                    },
                )?;

            self.client.send_notification::<Progress>(ProgressParams {
                token: NumberOrString::String(self.token.to_string()),
                value: ProgressParamsValue::WorkDone(WorkDoneProgress::Begin(
                    WorkDoneProgressBegin {
                        title: "Building".to_string(),
                        message: Some(uri.as_str().to_string()),
                        cancellable: Some(false),
                        percentage: None,
                    },
                )),
            })?;
        };
        Ok(())
    }
}

impl<'a> Drop for ProgressReporter<'a> {
    fn drop(&mut self) {
        if self.supports_progress {
            drop(self.client.send_notification::<Progress>(ProgressParams {
                token: NumberOrString::String(self.token.to_string()),
                value: ProgressParamsValue::WorkDone(WorkDoneProgress::End(WorkDoneProgressEnd {
                    message: None,
                })),
            }));
        }
    }
}

#[derive(Default)]
pub struct BuildEngine {
    lock: Mutex<()>,
    pub positions_by_uri: DashMap<Arc<Url>, Position>,
}

impl BuildEngine {
    pub fn build(
        &self,
        request: FeatureRequest<BuildParams>,
        client: LspClient,
    ) -> Result<BuildResult> {
        let lock = self.lock.lock().unwrap();

        let document = request
            .workspace
            .iter()
            .find(|document| {
                document
                    .data()
                    .as_latex()
                    .map_or(false, |data| data.extras.has_document_environment)
            })
            .unwrap_or_else(|| request.main_document());

        if !document.can_be_compiled() {
            log::info!(
                "Document {} cannot be compiled; skipping...",
                document.uri()
            );

            return Ok(BuildResult {
                status: BuildStatus::SUCCESS,
            });
        }

        if document.uri().scheme() != "file" {
            return Ok(BuildResult {
                status: BuildStatus::FAILURE,
            });
        }

        let path = document.uri().to_file_path().unwrap();

        let supports_progress = request
            .workspace
            .environment
            .client_capabilities
            .has_work_done_progress_support();

        let token = format!("texlab-build-{}", Uuid::new_v4());
        let progress_reporter = ProgressReporter {
            supports_progress,
            client: client.clone(),
            token: &token,
        };
        progress_reporter.start(document.uri())?;

        let options = &request.workspace.environment.options;

        let build_dir = options
            .root_directory
            .as_ref()
            .map(AsRef::as_ref)
            .or_else(|| path.parent())
            .unwrap();

        let args: Vec<_> = options
            .build
            .args
            .0
            .iter()
            .map(|arg| replace_placeholder(arg.clone(), &path))
            .collect();

        let mut process = Command::new(&options.build.executable.0)
            .args(args)
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .current_dir(build_dir)
            .spawn()?;

        let (exit_sender, exit_receiver) = crossbeam_channel::bounded(1);
        let log_handle = capture_output(&mut process, client, exit_receiver);
        let success = process.wait().map(|status| status.success())?;
        exit_sender.send(())?;
        drop(exit_sender);

        log_handle.join().unwrap();
        let status = if success {
            BuildStatus::SUCCESS
        } else {
            BuildStatus::ERROR
        };

        drop(progress_reporter);
        drop(lock);

        if let Some((executable, args)) = options
            .forward_search
            .executable
            .as_deref()
            .zip(options.forward_search.args.as_deref())
            .filter(|_| options.build.forward_search_after)
        {
            let position = self
                .positions_by_uri
                .get(&request.uri)
                .map(|entry| *entry.value())
                .unwrap_or_default();

            ForwardSearch::builder()
                .executable(executable)
                .args(args)
                .line(position.line)
                .workspace(&request.workspace)
                .tex_uri(&request.uri)
                .build()
                .execute();
        }

        Ok(BuildResult { status })
    }
}

fn capture_output(
    process: &mut std::process::Child,
    client: LspClient,
    exit_receiver: Receiver<()>,
) -> JoinHandle<()> {
    let (log_sender, log_receiver) = crossbeam_channel::unbounded();
    track_output(process.stdout.take().unwrap(), log_sender.clone());
    track_output(process.stderr.take().unwrap(), log_sender);
    thread::spawn(move || loop {
        crossbeam_channel::select! {
            recv(&log_receiver) -> message => {
                if let Ok(message) = message {
                    client.send_notification::<LogMessage>(
                        LogMessageParams {
                            message,
                            typ: lsp_types::MessageType::LOG,
                        },
                    )
                    .unwrap();
                }
            },
            recv(&exit_receiver) -> _ => break,
        };
    })
}

fn replace_placeholder(arg: String, file: &Path) -> String {
    if arg.starts_with('"') || arg.ends_with('"') {
        arg
    } else {
        arg.replace("%f", &file.to_string_lossy())
    }
}

fn track_output(output: impl Read + Send + 'static, sender: Sender<String>) -> JoinHandle<()> {
    let reader = BufReader::new(
        DecodeReaderBytesBuilder::new()
            .encoding(Some(encoding_rs::UTF_8))
            .utf8_passthru(true)
            .strip_bom(true)
            .build(output),
    );

    thread::spawn(move || {
        for line in reader.lines() {
            sender.send(line.unwrap()).unwrap();
        }
    })
}
