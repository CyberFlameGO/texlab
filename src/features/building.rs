mod progress;

use std::{
    io::{BufRead, BufReader, Read},
    path::Path,
    process::{Command, Stdio},
    thread::{self, JoinHandle},
};

use encoding_rs_io::DecodeReaderBytesBuilder;
use lsp_types::{notification::LogMessage, LogMessageParams, TextDocumentIdentifier, Url};
use serde::{Deserialize, Serialize};
use serde_repr::{Deserialize_repr, Serialize_repr};

use crate::{
    client::LspClient,
    db::{workspace::Workspace, Distro},
    ClientCapabilitiesExt, Db,
};

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BuildParams {
    pub text_document: TextDocumentIdentifier,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BuildResult {
    pub status: BuildStatus,
}

#[derive(Debug, PartialEq, Eq, Clone, Copy, Serialize_repr, Deserialize_repr)]
#[repr(i32)]
pub enum BuildStatus {
    SUCCESS = 0,
    ERROR = 1,
    FAILURE = 2,
    CANCELLED = 3,
}

#[derive(Debug)]
pub struct TexCompiler {
    uri: Url,
    progress: bool,
    command: Command,
    client: LspClient,
}

impl TexCompiler {
    pub fn configure(db: &dyn Db, uri: Url, client: LspClient) -> Option<Self> {
        let workspace = Workspace::get(db);
        let document = match workspace.lookup_uri(db, &uri) {
            Some(child) => workspace
                .parents(db, Distro::get(db), child)
                .iter()
                .next()
                .copied()
                .unwrap_or(child),
            None => return None,
        };

        if document.location(db).path(db).is_none() {
            log::warn!("Document {uri} cannot be compiled; skipping...");
            return None;
        }

        let options = &workspace.options(db).build;
        let mut command = Command::new(&options.executable.0);
        command
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .current_dir(
                workspace
                    .working_dir(db, document.location(db))
                    .path(db)
                    .as_deref()?,
            );

        let path = document.location(db).path(db).as_deref().unwrap();
        for arg in options.args.0.iter() {
            command.arg(replace_placeholder(arg, path));
        }

        Some(Self {
            uri: document.location(db).uri(db).clone(),
            progress: workspace
                .client_capabilities(db)
                .has_work_done_progress_support(),
            command,
            client,
        })
    }

    pub fn run(mut self) -> BuildStatus {
        let reporter = if self.progress {
            let inner = progress::Reporter::new(&self.client);
            inner.start(&self.uri).expect("report progress");
            Some(inner)
        } else {
            None
        };

        let mut process = match self.command.spawn() {
            Ok(process) => process,
            Err(_) => {
                log::error!("Failed to spawn process: {:?}", self.command.get_program());
                return BuildStatus::FAILURE;
            }
        };

        let (line_sender, line_receiver) = flume::unbounded();
        let (exit_sender, exit_receiver) = flume::unbounded();
        track_output(process.stderr.take().unwrap(), line_sender.clone());
        track_output(process.stdout.take().unwrap(), line_sender);
        let client = self.client.clone();
        let handle = std::thread::spawn(move || {
            let typ = lsp_types::MessageType::LOG;

            loop {
                let done = flume::Selector::new()
                    .recv(&line_receiver, |line| {
                        let message = line.unwrap();
                        client
                            .send_notification::<LogMessage>(LogMessageParams { message, typ })
                            .unwrap();
                        false
                    })
                    .recv(&exit_receiver, |_| true)
                    .wait();

                if done {
                    break;
                }
            }
        });

        let status = process.wait().map_or(BuildStatus::FAILURE, |result| {
            if result.success() {
                BuildStatus::SUCCESS
            } else {
                BuildStatus::ERROR
            }
        });

        exit_sender
            .send(())
            .expect("send exit signal to output reader");
        handle.join().unwrap();

        drop(reporter);
        status
    }
}

fn track_output(
    output: impl Read + Send + 'static,
    sender: flume::Sender<String>,
) -> JoinHandle<()> {
    let reader = BufReader::new(
        DecodeReaderBytesBuilder::new()
            .encoding(Some(encoding_rs::UTF_8))
            .utf8_passthru(true)
            .strip_bom(true)
            .build(output),
    );

    thread::spawn(move || {
        for line in reader.lines() {
            sender.send(line.unwrap()).unwrap();
        }
    })
}

fn replace_placeholder(arg: &str, file: &Path) -> String {
    if arg.starts_with('"') || arg.ends_with('"') {
        arg.to_string()
    } else {
        arg.replace("%f", &file.to_string_lossy())
    }
}
