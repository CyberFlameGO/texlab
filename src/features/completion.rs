mod acronym_ref;
mod argument;
mod begin_snippet;
pub mod builder;
mod citation;
mod color;
mod color_model;
mod component_command;
mod component_environment;
mod entry_type;
mod field;
mod glossary_ref;
mod import;
mod include;
mod label;
mod theorem;
mod tikz_library;
mod user_command;
mod user_environment;

use lsp_types::{CompletionList, Position, Url};

use crate::{features::completion::builder::CompletionBuilder, util::cursor::CursorContext, Db};

pub const COMPLETION_LIMIT: usize = 50;

pub fn complete(db: &dyn Db, uri: &Url, position: Position) -> Option<CompletionList> {
    let context = CursorContext::new(db, uri, position, ())?;
    let mut builder = CompletionBuilder::new(&context);
    log::debug!("[Completion] Cursor: {:?}", context.cursor);
    entry_type::complete_entry_types(&context, &mut builder);
    field::complete_fields(&context, &mut builder);
    argument::complete_arguments(&context, &mut builder);
    citation::complete_citations(&context, &mut builder);
    import::complete_imports(&context, &mut builder);
    color::complete_colors(&context, &mut builder);
    color_model::complete_color_models(&context, &mut builder);
    acronym_ref::complete_acronyms(&context, &mut builder);
    glossary_ref::complete_glossary_entries(&context, &mut builder);
    include::complete_includes(&context, &mut builder);
    label::complete_labels(&context, &mut builder);
    tikz_library::complete_tikz_libraries(&context, &mut builder);
    component_environment::complete_component_environments(&context, &mut builder);
    theorem::complete_theorem_environments(&context, &mut builder);
    user_environment::complete_user_environments(&context, &mut builder);
    begin_snippet::complete_begin_snippet(&context, &mut builder);
    component_command::complete_component_commands(&context, &mut builder);
    user_command::complete_user_commands(&context, &mut builder);
    Some(builder.finish())
}
