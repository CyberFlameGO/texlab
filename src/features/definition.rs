mod command;
mod document;
mod entry;
mod label;
mod string;

use std::sync::Arc;

use lsp_types::{GotoDefinitionParams, GotoDefinitionResponse, LocationLink, Url};
use rowan::TextRange;

use crate::LineIndexExt;

use self::{
    command::goto_command_definition, document::goto_document_definition,
    entry::goto_entry_definition, label::goto_label_definition, string::goto_string_definition,
};

use super::{cursor::CursorContext, FeatureRequest};

pub fn goto_definition(
    request: FeatureRequest<GotoDefinitionParams>,
) -> Option<GotoDefinitionResponse> {
    let context = CursorContext::new(request);
    log::debug!("[Definition] Cursor: {:?}", context.cursor);

    let origin_document = context.request.main_document();
    let links: Vec<_> = goto_command_definition(&context)
        .or_else(|| goto_document_definition(&context))
        .or_else(|| goto_entry_definition(&context))
        .or_else(|| goto_label_definition(&context))
        .or_else(|| goto_string_definition(&context))?
        .into_iter()
        .map(|result| {
            let origin_selection_range = Some(
                origin_document
                    .line_index
                    .line_col_lsp_range(result.origin_selection_range),
            );

            let target_document = &context.request.workspace.documents_by_uri[&result.target_uri];
            let target_uri = result.target_uri.as_ref().clone();
            let target_range = target_document
                .line_index
                .line_col_lsp_range(result.target_range);
            let target_selection_range = target_document
                .line_index
                .line_col_lsp_range(result.target_selection_range);

            LocationLink {
                origin_selection_range,
                target_uri,
                target_range,
                target_selection_range,
            }
        })
        .collect();

    Some(GotoDefinitionResponse::Link(links))
}

#[derive(Debug, Clone)]
struct DefinitionResult {
    origin_selection_range: TextRange,
    target_uri: Arc<Url>,
    target_range: TextRange,
    target_selection_range: TextRange,
}

#[cfg(test)]
mod tests;
