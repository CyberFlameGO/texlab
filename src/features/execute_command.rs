use std::{path::PathBuf, process::Stdio, sync::Arc};

use anyhow::Result;
use lsp_types::{TextDocumentIdentifier, Url};

use crate::Workspace;

pub fn execute_command(
    workspace: &Workspace,
    name: &str,
    args: Vec<serde_json::Value>,
) -> Result<()> {
    match name {
        "texlab.cleanAuxiliary" => {
            let params = args
                .into_iter()
                .next()
                .ok_or_else(|| anyhow::anyhow!("texlab.cleanAuxiliary requires one argument"))?;

            clean_output_files(workspace, CleanOptions::Auxiliary, params)?;
        }
        "texlab.cleanArtifacts" => {
            let params = args
                .into_iter()
                .next()
                .ok_or_else(|| anyhow::anyhow!("texlab.cleanArtifacts requires one argument"))?;

            clean_output_files(workspace, CleanOptions::Artifacts, params)?;
        }
        _ => anyhow::bail!("Unknown command: {}", name),
    }

    Ok(())
}

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Hash)]
enum CleanOptions {
    Auxiliary,
    Artifacts,
}

fn clean_output_files(
    workspace: &Workspace,
    options: CleanOptions,
    params: serde_json::Value,
) -> Result<()> {
    let params: TextDocumentIdentifier = serde_json::from_value(params)?;

    let uri = workspace
        .find_parent(&params.uri)
        .map(|document| document.uri)
        .unwrap_or_else(|| Arc::new(params.uri));

    if let Some(cx) = BuildContext::find(workspace, &uri) {
        let flag = match options {
            CleanOptions::Auxiliary => "-c",
            CleanOptions::Artifacts => "-C",
        };

        std::process::Command::new("latexmk")
            .arg(format!("-outdir={}", cx.output_dir.to_string_lossy()))
            .arg(flag)
            .arg(cx.input_file)
            .stdin(Stdio::null())
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status()?;
    }

    Ok(())
}

struct BuildContext {
    input_file: PathBuf,
    output_dir: PathBuf,
}

impl BuildContext {
    pub fn find(workspace: &Workspace, uri: &Url) -> Option<Self> {
        if uri.scheme() != "file" {
            return None;
        }

        let input_file = uri.to_file_path().ok()?;
        let options = &workspace.environment.options;
        let current_dir = &workspace.environment.current_directory;
        let output_dir = match (
            options.root_directory.as_ref(),
            options.aux_directory.as_ref(),
        ) {
            (_, Some(aux_dir)) => current_dir.join(aux_dir),
            (Some(root_dir), None) => current_dir.join(root_dir),
            (None, None) => input_file.parent()?.to_path_buf(),
        };

        log::info!("Output = {:#?}", output_dir);

        Some(Self {
            input_file,
            output_dir,
        })
    }
}
