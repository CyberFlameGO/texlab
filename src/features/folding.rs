use lsp_types::{FoldingRange, FoldingRangeKind, Range, Url};
use rowan::ast::AstNode;

use crate::{
    db::{parse::DocumentData, Workspace},
    syntax::{bibtex, latex},
    util::line_index_ext::LineIndexExt,
    Db,
};

pub fn find_all(db: &dyn Db, uri: &Url) -> Option<Vec<FoldingRange>> {
    let document = Workspace::get(db).lookup_uri(db, uri)?;
    let line_index = document.line_index(db);
    let foldings = match document.parse(db) {
        DocumentData::Tex(data) => {
            let mut results = Vec::new();
            let root = data.root(db);
            for node in root.descendants() {
                if let Some(folding) = latex::Environment::cast(node.clone())
                    .map(|node| latex::small_range(&node))
                    .or_else(|| {
                        latex::Section::cast(node.clone()).map(|node| latex::small_range(&node))
                    })
                    .or_else(|| latex::EnumItem::cast(node).map(|node| latex::small_range(&node)))
                    .map(|node| line_index.line_col_lsp_range(node))
                    .map(create_range)
                {
                    results.push(folding);
                }
            }

            results
        }
        DocumentData::Bib(data) => {
            let root = data.root(db);
            root.descendants()
                .filter(|node| {
                    matches!(
                        node.kind(),
                        bibtex::PREAMBLE | bibtex::STRING | bibtex::ENTRY
                    )
                })
                .map(|node| create_range(line_index.line_col_lsp_range(node.text_range())))
                .collect()
        }
        DocumentData::Log(_) | DocumentData::TexlabRoot(_) | DocumentData::Tectonic(_) => {
            return None;
        }
    };

    Some(foldings)
}

fn create_range(range: Range) -> FoldingRange {
    FoldingRange {
        start_line: range.start.line,
        start_character: Some(range.start.character),
        end_line: range.end.line,
        end_character: Some(range.end.character),
        collapsed_text: None,
        kind: Some(FoldingRangeKind::Region),
    }
}
