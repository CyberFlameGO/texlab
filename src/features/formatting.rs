mod bibtex_internal;
mod latexindent;

use lsp_types::{FormattingOptions, TextEdit, Url};

use crate::{
    db::{Language, Workspace},
    Db, Formatter,
};

use self::{bibtex_internal::format_bibtex_internal, latexindent::format_with_latexindent};

pub fn format_source_code(
    db: &dyn Db,
    uri: &Url,
    options: &FormattingOptions,
) -> Option<Vec<TextEdit>> {
    let workspace = Workspace::get(db);
    let document = workspace.lookup_uri(db, uri)?;
    let config = workspace.config(db);
    match document.language(db) {
        Language::Tex => match config.formatting.tex_formatter {
            Formatter::Null => None,
            Formatter::Server => None,
            Formatter::LatexIndent => format_with_latexindent(db, document),
        },
        Language::Bib => match config.formatting.bib_formatter {
            Formatter::Null => None,
            Formatter::Server => format_bibtex_internal(db, document, options),
            Formatter::LatexIndent => format_with_latexindent(db, document),
        },
        Language::Log | Language::TexlabRoot | Language::Tectonic => None,
    }
}
