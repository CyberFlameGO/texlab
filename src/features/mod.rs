mod build;
#[cfg(feature = "completion")]
mod completion;
mod cursor;
mod definition;
mod folding;
mod formatting;
mod forward_search;
mod highlight;
mod hover;
mod link;
mod lsp_kinds;
mod reference;
mod rename;
mod symbol;

use std::sync::Arc;

use crate::{Document, ServerContext, Workspace, WorkspaceSubset};

#[cfg(feature = "completion")]
pub use self::completion::{complete, CompletionItemData, COMPLETION_LIMIT};
pub use self::{
    build::{BuildEngine, BuildParams, BuildResult, BuildStatus},
    definition::goto_definition,
    folding::find_foldings,
    formatting::format_source_code,
    forward_search::{execute_forward_search, ForwardSearchResult, ForwardSearchStatus},
    highlight::find_document_highlights,
    hover::find_hover,
    link::find_document_links,
    reference::find_all_references,
    rename::{prepare_rename_all, rename_all},
    symbol::{find_document_symbols, find_workspace_symbols},
};

#[derive(Clone)]
pub struct FeatureRequest<P> {
    pub context: Arc<ServerContext>,
    pub params: P,
    pub workspace: Arc<dyn Workspace>,
    pub subset: WorkspaceSubset,
}

impl<P> FeatureRequest<P> {
    pub fn main_document(&self) -> &Document {
        &self.subset.documents[0]
    }
}

#[cfg(test)]
mod testing {
    use std::{path::PathBuf, sync::Arc};

    use lsp_types::{
        ClientCapabilities, ClientInfo, CompletionParams, DocumentFormattingParams,
        DocumentHighlightParams, DocumentLinkParams, FoldingRangeParams, FormattingOptions,
        GotoDefinitionParams, HoverParams, PartialResultParams, Position, ReferenceContext,
        ReferenceParams, RenameParams, TextDocumentIdentifier, TextDocumentPositionParams,
        WorkDoneProgressParams,
    };
    use typed_builder::TypedBuilder;

    use crate::{
        create_workspace_fast, distro::Resolver, DocumentLanguage, DocumentVisibility, Options,
        ServerContext, Uri, Workspace,
    };

    use super::*;

    #[derive(Debug, Clone, TypedBuilder)]
    pub struct FeatureTester<'a> {
        main: &'a str,

        files: Vec<(&'a str, &'a str)>,

        #[builder(default)]
        line: u32,

        #[builder(default)]
        character: u32,

        #[builder(default)]
        new_name: &'a str,

        #[builder(default)]
        include_declaration: bool,

        #[builder(default)]
        client_capabilities: ClientCapabilities,

        #[builder(default)]
        client_info: Option<ClientInfo>,

        #[builder(default)]
        resolver: Resolver,

        #[builder(default=std::env::temp_dir())]
        current_directory: PathBuf,

        #[builder(default, setter(strip_option))]
        root_directory: Option<PathBuf>,

        #[builder(default, setter(strip_option))]
        aux_directory: Option<PathBuf>,
    }

    impl<'a> FeatureTester<'a> {
        pub fn uri(&self, name: &str) -> Arc<Uri> {
            let path = self.current_directory.join(name);
            Arc::new(Uri::from_file_path(path).unwrap())
        }

        fn options(&self) -> Options {
            Options {
                aux_directory: self.aux_directory.clone(),
                root_directory: self.root_directory.clone(),
                ..Options::default()
            }
        }

        fn identifier(&self) -> TextDocumentIdentifier {
            let uri = self.uri(self.main);
            TextDocumentIdentifier::new(uri.as_ref().clone().into())
        }

        fn context(&self) -> Arc<ServerContext> {
            let cx = ServerContext::new(self.current_directory.clone());
            *cx.client_capabilities.lock().unwrap() = self.client_capabilities.clone();
            *cx.client_info.lock().unwrap() = self.client_info.clone();
            *cx.options.write().unwrap() = self.options();
            *cx.resolver.lock().unwrap() = self.resolver.clone();
            Arc::new(cx)
        }

        fn workspace(&self, cx: Arc<ServerContext>) -> Arc<dyn Workspace> {
            let workspace = create_workspace_fast(cx).unwrap();
            for (name, source_code) in &self.files {
                let uri = self.uri(name);
                let path = uri.to_file_path().unwrap();
                let language = DocumentLanguage::by_path(&path).expect("unknown document language");
                workspace.open(
                    uri,
                    Arc::new(source_code.trim().to_string()),
                    language,
                    DocumentVisibility::Visible,
                );
            }
            Arc::new(workspace)
        }

        fn request<P>(&self, params: P) -> FeatureRequest<P> {
            let cx = self.context();
            let workspace = self.workspace(Arc::clone(&cx));
            let subset = workspace.subset(self.uri(self.main)).unwrap();
            FeatureRequest {
                context: cx,
                workspace,
                subset,
                params,
            }
        }

        pub fn link(self) -> FeatureRequest<DocumentLinkParams> {
            let text_document = self.identifier();
            let params = DocumentLinkParams {
                text_document,
                work_done_progress_params: WorkDoneProgressParams::default(),
                partial_result_params: PartialResultParams::default(),
            };
            self.request(params)
        }

        pub fn folding(self) -> FeatureRequest<FoldingRangeParams> {
            let text_document = self.identifier();
            let params = FoldingRangeParams {
                text_document,
                work_done_progress_params: WorkDoneProgressParams::default(),
                partial_result_params: PartialResultParams::default(),
            };
            self.request(params)
        }

        pub fn reference(self) -> FeatureRequest<ReferenceParams> {
            let params = ReferenceParams {
                text_document_position: TextDocumentPositionParams::new(
                    self.identifier(),
                    Position::new(self.line, self.character),
                ),
                context: ReferenceContext {
                    include_declaration: self.include_declaration,
                },
                work_done_progress_params: WorkDoneProgressParams::default(),
                partial_result_params: PartialResultParams::default(),
            };
            self.request(params)
        }

        pub fn hover(self) -> FeatureRequest<HoverParams> {
            let params = HoverParams {
                text_document_position_params: TextDocumentPositionParams::new(
                    self.identifier(),
                    Position::new(self.line, self.character),
                ),
                work_done_progress_params: WorkDoneProgressParams::default(),
            };
            self.request(params)
        }

        pub fn completion(self) -> FeatureRequest<CompletionParams> {
            let params = CompletionParams {
                text_document_position: TextDocumentPositionParams::new(
                    self.identifier(),
                    Position::new(self.line, self.character),
                ),
                work_done_progress_params: WorkDoneProgressParams::default(),
                partial_result_params: PartialResultParams::default(),
                context: None,
            };

            self.request(params)
        }

        pub fn definition(self) -> FeatureRequest<GotoDefinitionParams> {
            let params = GotoDefinitionParams {
                text_document_position_params: TextDocumentPositionParams::new(
                    self.identifier(),
                    Position::new(self.line, self.character),
                ),
                work_done_progress_params: WorkDoneProgressParams::default(),
                partial_result_params: PartialResultParams::default(),
            };
            self.request(params)
        }

        pub fn rename(self) -> FeatureRequest<RenameParams> {
            let params = RenameParams {
                text_document_position: TextDocumentPositionParams::new(
                    self.identifier(),
                    Position::new(self.line, self.character),
                ),
                new_name: self.new_name.to_string(),
                work_done_progress_params: WorkDoneProgressParams::default(),
            };
            self.request(params)
        }

        pub fn formatting(self) -> FeatureRequest<DocumentFormattingParams> {
            let params = DocumentFormattingParams {
                text_document: self.identifier(),
                work_done_progress_params: WorkDoneProgressParams::default(),
                options: FormattingOptions::default(),
            };
            self.request(params)
        }

        pub fn highlight(self) -> FeatureRequest<DocumentHighlightParams> {
            let params = DocumentHighlightParams {
                text_document_position_params: TextDocumentPositionParams::new(
                    self.identifier(),
                    Position::new(self.line, self.character),
                ),
                work_done_progress_params: WorkDoneProgressParams::default(),
                partial_result_params: PartialResultParams::default(),
            };
            self.request(params)
        }
    }
}
