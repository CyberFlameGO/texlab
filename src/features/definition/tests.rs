use insta::{
    assert_json_snapshot,
    internals::{Content, ContentPath},
};
use lsp_types::Url;

use crate::features::{goto_definition, testing::FeatureTester};

#[test]
fn test_empty_latex_document() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![("main.tex", "")])
            .main("main.tex")
            .line(0)
            .character(0)
            .build()
            .definition(),
    );

    assert_eq!(links, None);
}

#[test]
fn test_empty_bibtex_document() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![("main.bib", "")])
            .main("main.bib")
            .line(0)
            .character(0)
            .build()
            .definition(),
    );

    assert_eq!(links, None);
}

#[test]
fn test_command_definition() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![(
                "main.tex",
                r#"
\DeclareMathOperator{\foo}{foo}
\foo"#,
            )])
            .main("main.tex")
            .line(1)
            .character(2)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_document_simple() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![
                ("foo.tex", r#"\addbibresource{baz.bib}"#),
                ("bar.bib", r#"@article{foo, bar = {baz}}"#),
                ("baz.bib", r#"@article{foo, bar = {baz}}"#),
            ])
            .main("foo.tex")
            .line(0)
            .character(18)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_entry_simple() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![
                (
                    "foo.tex",
                    r#"
\addbibresource{baz.bib}
\cite{foo}"#,
                ),
                ("bar.bib", r#"@article{foo, bar = {baz}}"#),
                ("baz.bib", r#"@article{foo, bar = {baz}}"#),
            ])
            .main("foo.tex")
            .line(1)
            .character(6)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_string_simple() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = {bar}}
@article{bar, author = foo}"#,
            )])
            .main("main.bib")
            .line(1)
            .character(24)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_string_concat() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = {bar}}
@article{bar, author = foo # "bar"}"#,
            )])
            .main("main.bib")
            .line(1)
            .character(24)
            .build()
            .definition(),
    );

    assert_json_snapshot!( links, { "[].targetUri" => insta::dynamic_redaction(redact_uri) });
}

#[test]
fn test_string_field() {
    let links = goto_definition(
        FeatureTester::builder()
            .files(vec![(
                "main.bib",
                r#"
@string{foo = {bar}}
@article{bar, author = foo}"#,
            )])
            .main("main.bib")
            .line(1)
            .character(18)
            .build()
            .definition(),
    );

    assert_eq!(links, None);
}

fn redact_uri(value: Content, _path: ContentPath) -> String {
    value.as_str().unwrap().replace(
        Url::from_directory_path(std::env::temp_dir())
            .unwrap()
            .as_str(),
        "[tmp]/",
    )
}
