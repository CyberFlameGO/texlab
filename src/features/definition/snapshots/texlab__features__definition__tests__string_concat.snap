---
source: src/features/definition/tests.rs
assertion_line: 87
expression: links
---
[
  {
    "originSelectionRange": {
      "start": {
        "line": 1,
        "character": 23
      },
      "end": {
        "line": 1,
        "character": 26
      }
    },
    "targetUri": "[tmp]/main.bib",
    "targetRange": {
      "start": {
        "line": 0,
        "character": 0
      },
      "end": {
        "line": 0,
        "character": 20
      }
    },
    "targetSelectionRange": {
      "start": {
        "line": 0,
        "character": 8
      },
      "end": {
        "line": 0,
        "character": 11
      }
    }
  }
]
